File: Hackathon/modules/Java/DummyNumberRecogniser.java
import java.io.*;
import java.util.*;
import org.json.simple.*;
import org.json.simple.parser.*;

public class DummyNumberRecogniser
{
	public static void main(String[] args) throws IOException
	{
		try (BufferedReader br = new BufferedReader(new InputStreamReader(System.in)))
		{
			String line;
			while ((line = br.readLine()) != null)
			{
				Object obj = JSONValue.parse(line);
				if (obj instanceof JSONObject)
				{
					process((JSONObject)obj);
				}
			}
		}
		System.err.println("dummy number recogniser java is done!");
	}
	
	@SuppressWarnings("unchecked")
	static void process(JSONObject data)
	{
		int frameSum = 0;
		List numberLocations = (List)data.get("numberLocations");
		for (Object numberLocation : numberLocations)
		{
			int frameNr = ((Number)((Map)numberLocation).get("frameNr")).intValue();
			frameSum += frameNr;
		}
		float averageFrame = frameSum / (float)numberLocations.size();
		int digits = (int)(averageFrame * 218); // Just some random multiplication to get more digits
		int checksum = generateUICChecksum(digits);
		String uic = String.format("%011d-%d", digits, checksum);

		JSONObject outData = new JSONObject();
		outData.put("wagonStart", data.get("wagonStart"));
		outData.put("number", uic);

		System.out.println(outData);
	}

	static int generateUICChecksum(long digits)
	{
		boolean evenPosition = false;
		int sum = 0;
		while (digits > 0)
		{
			int factor = evenPosition ? 1 : 2;
			int digit = (digits % 10) * factor;
			while (digit > 0)
			{
				sum += digit % 10;
				digit /= 10;
			}
			digits /= 10;
			evenPosition = !evenPosition;
		}
		int roundUp = 10 * (int)Math.ceil(sum / 10.0f);
		return roundUp - sum;
	}
}

File: Hackathon/modules/Python/dummyNumberRecogniser.py
#!/usr/bin/python

import sys
import json
import math

def generateUICChecksum(digits):
	digits = int(digits)
	evenPosition = False
	sum = 0
	while digits > 0:
		factor = 1 if evenPosition else 2
		digit = (digits % 10) * factor
		while digit > 0:
			sum += digit % 10
			digit = int(digit / 10)
		digits = int(digits / 10)
		evenPosition = not evenPosition
	roundUp = math.ceil(sum / 10.0) * 10
	return roundUp - sum

while True:
	line = sys.stdin.readline()
	if not line:
		break
	data = json.loads(line)

	frameSum = 0
	for numberLocation in data["numberLocations"]:
		frameSum = frameSum + int(numberLocation["frameNr"])
	averageFrame = frameSum / float(len(data["numberLocations"]))
	digits = averageFrame * 218 # This is just to get some interesting digits
	checksum = generateUICChecksum(digits)
	uic = "%011d-%d" % (digits, checksum)

	outData = {"wagonStart": data["wagonStart"], "number": uic}
	outLine = json.dumps(outData)
	print(outLine)

sys.stderr.write("Dummy number recogniser python is done!\n")